[inode.c] Don't fetch THIS unconditionally

black-dragon74 opened the pull request:
In some cases we can skip fetching `THIS` as we return before
we consume its value. This patch optimizes such cases.

Signed-off-by: black.dragon74 <nickk.2974@gmail.com>

black-dragon74 committed:
[inode.c] Don't fetch THIS unconditionally

In some cases we can skip fetching THIS as we return before
we consume its value. This patch optimizes such cases.

Signed-off-by: black.dragon74 <nickk.2974@gmail.com>

black-dragon74 committed:
[inode.c] We can skip checking for root GFID in some cases

Signed-off-by: black.dragon74 <nickk.2974@gmail.com>

mykaul commented on libglusterfs/src/inode.c:
```diff
@@ -580,7 +577,7 @@ __inode_ref(inode_t *inode, bool is_invalidate)
      * count as 1 always
      */
-    if (__is_root_gfid(inode->gfid) && inode->ref)
+    if (inode->ref && __is_root_gfid(inode->gfid))
```
nit (in a different patch!) we can do the is root gfid outside this code path, to the best of my knowledge.

black-dragon74 replied:
IIRC, Xavi mentioned that doing so (moving root GFID checks outside `__inode_red/unref`) would require the caller to always remember doing them checks manually and one might forget doing so. In that case, we may actually release and destroy the root inode quite easily which could prove dangerous.

xhernandez replied:
Yes. That's what I said. However now I've realized that `__inode_ref()` and `_inode_unref()` are declared `static`, which I missed in the other review, so their scope is very limited and we could move the check to the callers from this same module. In this case it's under control. I'll update the other PR.

black-dragon74 commented:
/run regression

xhernandez commented on libglusterfs/src/inode.c:
```diff
@@ -541,6 +539,7 @@ __inode_unref(inode_t *inode, bool clear)
     --inode->ref;
+    this = THIS;
 
     index = __inode_get_xl_index(inode, this);
     if (index >= 0) {
```
Do we really need to do this ? it could be useful for debugging in some cases, though I've never used it, but requiring `THIS` always just for a debugging purpose seems too costly.

black-dragon74 replied:
IIRC, we had a discussion about this in some other patches as well. TBH, I do not know what is the right step to take here. Maybe you and @mykaul can decide on if it is really needed else we can always drop this.

xhernandez replied:
@mohit84 @amarts @pranithk do you think it's useful/necessary to keep `xl_key` and `ref` into the ctx array ? IMO we could remove them or at least only keep them for debug builds

amarts replied:
Yes, I am fine with keeping them only in DEBUG builds. Production builds needn't have it IMO.

black-dragon74 committed:
Removed conflicts

Let us keep this patch only for once purpose, i.e. optimizing fetching
`THIS`. The other change is already taken care of by #3202

Signed-off-by: black-dragon74 <niryadav@redhat.com>

black-dragon74 commented:
Also, wrt to #3202 this patch is obsolete.

black-dragon74 commented:
/recheck smoke
